### PR TITLE
Allow binding on the un predetermined port for the callback server

### PR DESCRIPTION
**Pull request by 243826**
As discussed in #147 -  and do late initialization of the callback client.

### CODE REVIEW

**Comment by 243826** [comment]
@bartdag attention please :-)

File: py4j-python/src/py4j/java_gateway.py [hunk 1]
```diff
@@ -699,7 +699,7 @@ def start(self):
             self.stream = self.socket.makefile("rb", 0)
         except Exception as e:
             msg = "An error occurred while trying to connect to the Java "\
-                "server"
+                "server (%s:%s)" % (self.address, self.port)
```
Comment (bartdag) [inline py4j-python/src/py4j/java_gateway.py]: we use the .format version of string formatting :-)

File: py4j-python/src/py4j/java_gateway.py [hunk 2]
```diff
@@ -1598,8 +1601,9 @@ def start(self):
             socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         try:
             self.server_socket.bind((self.address, self.port))
+            self.__listening_address, self.__listening_port = self.server_socket.getsockname()
         except Exception as e:
-            msg = "An error occurred while trying to start the callback server"
+            msg = "An error occurred while trying to start the callback server (%s:%d)" % (self.address, self.port)
```
Comment (bartdag) [inline py4j-python/src/py4j/java_gateway.py]: this is a very good idea. I should print these info as often as possible. Please use .format though.

File: py4j-python/src/py4j/java_gateway.py [hunk 3]
```diff
@@ -1598,8 +1601,9 @@ def start(self):
             socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
         try:
             self.server_socket.bind((self.address, self.port))
+            self.__listening_address, self.__listening_port = self.server_socket.getsockname()
```
Comment (bartdag) [inline py4j-python/src/py4j/java_gateway.py]: only one underscore (no magic attribute or name mangling), e.g., ``_listening_address``

**Comment by bartdag** [comment]
Thanks for your pull request. This provides greater flexibility and visibility and I like it.

I need to think a little about the CallbackClient class.

1. I like that you can specify the address and the port of the CallbackClient after the GatewayServer is created. It makes sense.

2. I don't like the mutability of GatewayClient because I don't want the port and address to change after call back connections are created. I think this should be handled in GatewayServer, something like setupNewGatewayClient(address, port), which would a) stop and cleanup the existing gateway client and b) creates a new gateway client with immutable port and address. Not sure about the setupNewGatewayClient though.

What do you think?

**Comment by 243826** [comment]
Thanks for your review and appreciate it. I'll make the changes related to format and mangling and submit the pull request as you suggested.

I agree with you that setupNewGatewayClient as it seems cleaner as it clearly communicates the intention and gives an opportunity to do a cleanup. I would call it changePythonServerAddress(iface, port) or changePythonAddress(iface, port). But anyOtherName is good as well if you feel strongly about it.

Rationale:
May be it's just me that it took some time for me to wrap my head around clients/server concept  in py4j and how they map to Python side and JVM side. In python process - I expected GatewayServer would be something that python process launches and from JVM process jvm process would do the same. But the GatewayServer and CallbackServer meanings do not change and as a new comer I had to learn it by experimenting with ports. It's not that big of an issue but to make it super clear, I would have called them JVMServer and PythonServer instead of GatewayServer and CallbackServer respectively.

Let me know what your thoughts are and I may be able to code it up and update the pull request.

**Comment by bartdag** [comment]
> I would have called them JVMServer and PythonServer instead of GatewayServer and CallbackServer respectively.

This is mainly due to evolution. At first, only one direction was supported (Java to Python) so this distinction was not required. I like the names you propose and may reuse them for the new threading model I will be working on next year.

**Comment by 243826** [comment]
I saw that my pull request failed the CI. I did not think that I changed anything to cause it. Will investigate and resubmit it. Thanks.

**Comment by 243826** [comment]
@bartdag I am trying to get to the root of the problems with the tests failing w/ travis CI and noticed that although the py4j-java is actually coded using maven, you are using ant to compile it.

The downside of using ant is dependency resolution which I think is forte of maven. If maven is used to test the project with travis, you would not need to bundle and ship junit jar file as you are doing currently now. The dependency resolution happens at compile/test time and is different for the 2 phases. So I would recommend against using ant.

Do you like ant for some reason (other than probably legacy)?

**Comment by 243826** [comment]
Hi @bartdag, I spent some time today looking at the intermittent test failures. It looks like the race condition within and amongst the various versions of test (total 6 runs - python versions and java versions running together). So the travis CI failures can be written off.

I want to help fix this issue also but would love your buy in and support.

**Comment by 243826** [comment]
Also - would really appreciate if you make a release with the changes I submitted so that I can evangelize py4j. Otherwise I am forced to make a local release which many do not appreciate.

**Comment by bartdag** [comment]
Hi, sorry for the delays, I'm currently very busy with work and life in general with the coming holidays :-)

Thanks for your work on Py4J, I'll try to be more responsive.

Regarding this pull request:

1. The tests are indeed tricky and if they are green locally, it's fine with me. I've been trying to improve the test suite's reliability but it is quite challenging.

2. I believe the code still need to be refactored. I would use the name resetCallbackClientAddress, which conveys that there is a reset going on. I do not want to introduce a new terminology (i.e., PythonServer) for now, but I will seriously take this name into consideration when I get rid of the "Gateway/Callback" terminology, hopefully in the near future.

3. Making a release is time consuming, especially deploying to maven central. I'll try to make one next week, but I cannot promise anything.

4. Regarding the use of ant: I went through a few builders (paver, then ant, then someone asked for maven support, etc.). I'm not a huge fan of maven because it seems overly complex for it's purpose and I would much prefer to move to gradle instead of spending time to convert the whole release process from ant to maven. I totally agree that including the junit jar is more a hack and is not necessary with more modern build tools :-)

**Comment by 243826** [comment]
@bartdag See my latest commit on this pull request. I think this is the refactoring you wanted to see?

**Comment by bartdag** [comment]
Hi, yes, this looks good. The errors reported by travis are weird though, but it might be time to update to update the travis config for their new build system or to move to circle-ci. As soon as I can try it locally I'll merge the pull request and start to work on a release (there are a few fixes in @master that require a few bits of documentation).

Thanks a lot for your work and your input!

**Comment by 243826** [comment]
Thanks!

**Comment by 243826** [comment]
@bartdag - I should have tested thoroughly. The resetCallbackClientAddress does not work. It seems that the callback client reference is deeply distributed with the proxies. With the resetCallbackClientAddress, we will need to recall all the references to  CallbackClient being shutdown and it seems complicated. My last commit I added yesterday does not work.

Sorry.

**Comment by bartdag** [comment]
Well my fault for not trying more the feature. Proxies that were started before the reset can fail (similar behaviour to Java objects on the Python side that are accessed after a gateway is shut down). But the thing I did not check is if all sockets are correctly closed during the reset.

**Comment by 243826** [comment]
Submitted PR #183.

**Comment by 243826** [comment]
BTW, slightly orthogonal, I have a hunch that the java tests fail because they all start on the same machine concurrently and compete for the same ports. On my machine I do serial runs and have yet to see one fail so far (40+).

**Comment by bartdag** [comment]
Travis CI recently changed (again!!) their testing infrastructure and I believe you are right. That's also why I'm looking into circleci. It's slightly less convenient to run these kinds of matrix builds but our experience at work has been more positive than travis.

**Pull request merged**